CRM/Civisocial/OAuthProvider.php

```diff
@@ -0,0 +1,263 @@
+
+  /**
+   * HTTP Status code of last API request
+   *
+   * @var string
+   */
+  protected $httpCode;
+
+  /**
+   * Default timeout
+   *
+   * @var int
+   */
+  protected $timeout = 30;
+
+  /**
+   * Default connection timeout
+   *
+   * @var int
+   */
+  protected $connectTimeout = 30;
+
+  /**
+   * Verify SSL certificate
+   *
+   * @var bool
+   */
+  protected $sslVerifyPeer = FALSE;
+
+  /**
+   * Contains the last HTTP headers returned
+   *
+   * @var string
+   */
+  protected $httpInfo = array();
+
+  /**
+   * Get social user information
+   *
+   * @return array
+   */
+  public function getUserProfile() {
+    return $this->userProfile;
+  }
+
+  /**
+   * Retrieve API credentails for the given Social Media
+   *
+   * @param string $backend
+   *      OAuth Provider short name (alias)
+   */
+  public function getApiCredentials($backend) {
+    $this->apiKey = civicrm_api3(
+            "setting",
+            "getvalue",
+            array(
+              "group" => "CiviSocial Account Credentials",
+              "name" => "{$backend}_api_key",
+            )
+        );
+    $this->apiSecret = civicrm_api3(
+            "setting",
+            "getvalue",
+            array(
+              "group" => "CiviSocial Account Credentials",
+              "name" => "{$backend}_api_secret",
+            )
+        );
+  }
+
+  /**
+   * URL to be redirected to after user authorizes
+   *
+   * @param string $backend
+   *      OAuth Provider short name (alias)
+   *
+   * @return string
+   */
+  public function getCallbackUri($backend) {
+    return rawurldecode(CRM_Utils_System::url("civicrm/civisocial/callback/{$backend}", NULL, TRUE));
+  }
+
+  /**
+   * Authorization URI that user will be redirected to for login
+   */
+  public function getLoginUri() {
+  }
+
+  /**
+   * Process information returned by OAuth provider after login
+   */
+  public function handleCallback() {
+    if ($this->isLoggedIn()) {
+      CRM_Utils_System::redirect(CRM_Utils_System::url('', NULL, TRUE));
+    }
+  }
+
+  /**
+   * Save Backend information to the session
+   *
+   * @param string $backend
+   *      Shortname for OAuth provider
+   * @param string $accessToken
+   *      Access Token provided by OAuth provider after successfull authentication
+   * @param string $backendId
+   *      Unique user ID to OAuthProvider
+   */
+  public function login($backend, $accessToken, $backendId) {
+    $session = CRM_Core_Session::singleton();
+    $session->set('civisocial_logged_in', TRUE);
+    $session->set('civisocial_backend', $backend);
+    $session->set('civisocial_backend_user_id', $backendId);
+    $session->set('access_token', $accessToken);
+  }
+
+  /**
+   * Disconnect with OAuth provider
+   */
+  public function logout() {
+    $session = CRM_Core_Session::singleton();
+    $session->set('civisocial_logged_in', NULL);
+    $session->set('civisocial_backend', NULL);
+    $session->set('access_token', NULL);
+  }
+
```

Comment:
@pradpnayak We can. And actually, I was looking for a way to unset those session values than simply setting them NULL. `unset($_SESSION['CiviCRM']['access_token']` would kinda break the pattern. And Session class doesn't seem to have a way to unset. Or unsetting is not necessary?